File: frontend/src/api/rca.ts
import { apiClient } from "@/api/client";

type ApiEnvelope<T> = {
  code: number;
  message: string;
  data: T;
};

export type RCARun = {
  id: number;
  userId: number;
  conversationId?: number;
  incidentId?: number;
  workflowRunId?: number;
  status: string;
  query: string;
  scope: Record<string, unknown>;
  currentRound: number;
  maxRounds: number;
  timeoutAt?: string;
  cancelRequestedAt?: string;
  errorCode?: string;
  errorMessage?: string;
  stopReason?: string;
  startedAt?: string;
  finishedAt?: string;
  createdAt: string;
  updatedAt: string;
};

export type RCAHypothesis = {
  id?: string;
  summary: string;
  confidence: number;
  evidenceIds: number[];
};

export type RCANextAction = {
  actionKey: string;
  skillName: string;
  input: Record<string, unknown>;
};

export type RCARound = {
  id: number;
  runId: number;
  roundNumber: number;
  status: string;
  inputHypotheses: RCAHypothesis[];
  newEvidenceIds: number[];
  rejectedHypotheses: RCAHypothesis[];
  nextActions: RCANextAction[];
  errorCode?: string;
  startedAt?: string;
  finishedAt?: string;
};

export type RCAAction = {
  id: number;
  runId: number;
  roundId: number;
  actionKey: string;
  skillName: string;
  status: string;
  input: Record<string, unknown>;
  output?: unknown;
  evidenceIds: number[];
  errorCode?: string;
  errorMessage?: string;
  sensitiveRead: boolean;
  attempt: number;
  startedAt?: string;
  finishedAt?: string;
};

export type RCAEvidence = {
  id: number;
  evidenceKey: string;
  sourceType: string;
  sourceRef?: Record<string, unknown>;
  observedAt?: string;
  title?: string;
  summary: string;
  content?: unknown;
  confidence?: number;
  sensitivity?: string;
  rcaRunId?: number;
  rcaRoundId?: number;
  rcaActionId?: number;
  evidenceKind?: string;
  entity?: Record<string, unknown>;
  windowStart?: string;
  windowEnd?: string;
  sourceSkill?: string;
  dataSourceId?: number;
  createdAt: string;
};

export type RCARootCauseCandidateRecord = {
  id: number;
  runId: number;
  roundId: number;
  summary: string;
  confidence: number;
  evidenceIds: number[];
  rejected: boolean;
};

export type RCADetail = {
  run: RCARun;
  rounds: RCARound[];
  actions: RCAAction[];
  evidence: RCAEvidence[];
  rootCauseCandidates: RCARootCauseCandidateRecord[];
};

export type RCAPlannerHypothesis = {
  id: string;
  summary: string;
  confidence: number;
  supportingEvidenceIds: number[];
  contradictingEvidenceIds: number[];
  rationale?: string;
};

export type RCAPlannerAction = {
  actionKey: string;
  skillName: string;
  input: Record<string, unknown>;
  reason: string;
  targetEntity?: string;
  evidenceIds?: number[];
};

export type RCAPlannerResult = {
  hypotheses: RCAPlannerHypothesis[];
  missingEvidence: string[];
  nextActions: RCAPlannerAction[];
  shouldStop: boolean;
  stopReason: string;
  plannerDegraded: boolean;
  degradationReason?: string;
};

export type RCATopologyCandidate = {
  nodeKey: string;
  name: string;
  kind: string;
  componentType: string;
  sourceType: string;
  hops: number;
  edgeType?: string;
  direction: string;
  confidence: number;
  score: number;
  freshness: string;
  aliasMatched: boolean;
  conflict: boolean;
  selected: boolean;
  dataSourceId?: number;
  bindingStatus?: string;
  topologyEvidenceIds: number[];
};

export type RCATopologyInvestigation = {
  rootNodeKey?: string;
  observedAliases: string[];
  candidates: RCATopologyCandidate[];
  selected: RCATopologyCandidate[];
  missingEvidence: string[];
  conflicts: string[];
  fallbackUsed: boolean;
};

export type RCADatabaseDiagnosis = {
  provider: string;
  sourceType: string;
  dataSourceId: number;
  serviceName?: string;
  environment?: string;
  windowMinutes: number;
  correlationDimensions: string[];
  sqlFingerprint?: string;
  sanitizedSql?: string;
  missingEvidence: string[];
  supportingEvidenceIds: number[];
  assessment?: {
    status: string;
    highestImpactFingerprint?: string;
    categories: Array<{
      category: string;
      sourceSkill: string;
      collected: boolean;
      evidenceIds: number[];
    }>;
    evidenceIds: number[];
    missingEvidence: string[];
    rootCauseEligible: boolean;
    confidence: string;
    conclusion: string;
  };
};

export type RCAOrchestratorRound = {
  roundNumber: number;
  status: string;
  plan?: RCAPlannerResult;
  topologyInvestigation?: RCATopologyInvestigation;
  databaseDiagnosis?: RCADatabaseDiagnosis;
  actionIds: number[];
  evidenceIds: number[];
  errors?: string[];
};

export type RCAReportEvidenceRef = {
  id: number;
  evidenceKey: string;
  kind: string;
  sourceType: string;
  sourceSkill?: string;
  summary: string;
  confidence?: number;
  roundId?: number;
  actionId?: number;
  dataSourceId?: number;
  url: string;
};

export type RCAReport = {
  version: string;
  runId: number;
  status: string;
  query: string;
  scope: Record<string, unknown>;
  impactScope: {
    serviceName?: string;
    environment?: string;
    namespace?: string;
    windowStart?: string;
    windowEnd?: string;
    entities: string[];
  };
  evidence: {
    facts: RCAReportEvidenceRef[];
    rules: RCAReportEvidenceRef[];
    knowledge: RCAReportEvidenceRef[];
    hypotheses: RCAReportEvidenceRef[];
  };
  rootCauseCandidates: Array<{
    id: number;
    summary: string;
    status: string;
    confidence: number;
    evidenceStrength: string;
    evidenceStrengthScore: number;
    supportingEvidence: RCAReportEvidenceRef[];
    contradictingEvidence: RCAReportEvidenceRef[];
  }>;
  rejectedHypotheses: Array<{
    id?: string;
    summary: string;
    confidence: number;
    evidence: RCAReportEvidenceRef[];
    status: string;
  }>;
  investigation: Array<{
    roundNumber: number;
    status: string;
    checked: Array<{
      actionId: number;
      actionKey: string;
      skillName: string;
      status: string;
      evidenceIds: number[];
      errorCode?: string;
    }>;
    findings: RCAReportEvidenceRef[];
    continueReason?: string;
    stopReason?: string;
  }>;
  missingEvidence: string[];
  incomplete: boolean;
  conclusion: string;
  suggestions: Array<{
    summary: string;
    evidenceIds: number[];
    advisoryOnly: boolean;
    autoExecute: boolean;
  }>;
  riskNotices: string[];
  stopReason: string;
};

export type RCAOrchestratorResult = {
  version: string;
  run: RCARun;
  report?: RCAReport;
  stopReason: string;
  rounds: RCAOrchestratorRound[];
  degraded: boolean;
};

export async function createRCARun(input: {
  query: string;
  scope: Record<string, unknown>;
  maxRounds?: number;
  timeoutSeconds?: number;
}) {
  const response = await apiClient.post<ApiEnvelope<RCARun>>(
    "/api/rca/runs",
    input,
  );
  return response.data.data;
}

export async function listRCARuns(limit = 20) {
  const response = await apiClient.get<ApiEnvelope<RCARun[]>>("/api/rca/runs", {
    params: { limit },
  });
  return response.data.data;
}

export async function getRCADetail(runId: number) {
  const response = await apiClient.get<ApiEnvelope<RCADetail>>(
    `/api/rca/runs/${runId}`,
  );
  return response.data.data;
}

export async function orchestrateRCARun(runId: number) {
  const response = await apiClient.post<ApiEnvelope<RCAOrchestratorResult>>(
    `/api/rca/runs/${runId}/orchestrate`,
    {
      budget: {
        maxRounds: 3,
        maxSkillCallsPerRound: 12,
        maxSkillCalls: 24,
        maxConcurrentSkills: 4,
        maxTokens: 16000,
        maxContextBytes: 65536,
        maxWallTimeSeconds: 300,
        confidenceThreshold: 0.85,
      },
      useLlm: true,
    },
  );
  return response.data.data;
}

export async function cancelRCARun(runId: number) {
  const response = await apiClient.post<ApiEnvelope<RCARun>>(
    `/api/rca/runs/${runId}/cancel`,
  );
  return response.data.data;
}

export async function recoverRCARun(runId: number) {
  const response = await apiClient.post<
    ApiEnvelope<{
      run: RCARun;
      skippedActionIds: number[];
      retryableActionIds: number[];
    }>
  >(`/api/rca/runs/${runId}/recover`);
  return response.data.data;
}

export async function getRCAReport(runId: number) {
  const response = await apiClient.get<ApiEnvelope<RCAReport>>(
    `/api/rca/runs/${runId}/report`,
  );
  return response.data.data;
}

File: backend/internal/model/rca.go
package model

import (
	"encoding/json"
	"time"
)

const (
	RCARunStatusPending        = "pending"
	RCARunStatusRunning        = "running"
	RCARunStatusPartialSuccess = "partial_success"
	RCARunStatusSuccess        = "success"
	RCARunStatusFailed         = "failed"
	RCARunStatusCancelled      = "cancelled"
	RCARunStatusTimedOut       = "timed_out"

	RCARoundStatusPending        = "pending"
	RCARoundStatusRunning        = "running"
	RCARoundStatusPartialSuccess = "partial_success"
	RCARoundStatusSuccess        = "success"
	RCARoundStatusFailed         = "failed"
	RCARoundStatusCancelled      = "cancelled"
	RCARoundStatusTimedOut       = "timed_out"

	RCAActionStatusPending        = "pending"
	RCAActionStatusRunning        = "running"
	RCAActionStatusPartialSuccess = "partial_success"
	RCAActionStatusSuccess        = "success"
	RCAActionStatusFailed         = "failed"
	RCAActionStatusSkipped        = "skipped"
	RCAActionStatusCancelled      = "cancelled"
	RCAActionStatusTimedOut       = "timed_out"
)

type RCARun struct {
	ID                int64           `gorm:"column:id;primaryKey" json:"id"`
	UserID            int64           `gorm:"column:user_id;not null" json:"userId"`
	ConversationID    *int64          `gorm:"column:conversation_id" json:"conversationId,omitempty"`
	IncidentID        *int64          `gorm:"column:incident_id" json:"incidentId,omitempty"`
	WorkflowRunID     *int64          `gorm:"column:workflow_run_id" json:"workflowRunId,omitempty"`
	Status            string          `gorm:"column:status;size:30;not null" json:"status"`
	Query             string          `gorm:"column:query;not null" json:"query"`
	Scope             json.RawMessage `gorm:"column:scope;type:jsonb;not null" json:"scope"`
	CurrentRound      int             `gorm:"column:current_round;not null" json:"currentRound"`
	MaxRounds         int             `gorm:"column:max_rounds;not null" json:"maxRounds"`
	TimeoutAt         *time.Time      `gorm:"column:timeout_at" json:"timeoutAt,omitempty"`
	CancelRequestedAt *time.Time      `gorm:"column:cancel_requested_at" json:"cancelRequestedAt,omitempty"`
	ErrorCode         *string         `gorm:"column:error_code;size:80" json:"errorCode,omitempty"`
	ErrorMessage      *string         `gorm:"column:error_message" json:"errorMessage,omitempty"`
	StartedAt         *time.Time      `gorm:"column:started_at" json:"startedAt,omitempty"`
	FinishedAt        *time.Time      `gorm:"column:finished_at" json:"finishedAt,omitempty"`
	CreatedAt         time.Time       `gorm:"column:created_at;autoCreateTime" json:"createdAt"`
	UpdatedAt         time.Time       `gorm:"column:updated_at;autoUpdateTime" json:"updatedAt"`
}

func (RCARun) TableName() string { return "rca_run" }

type RCARound struct {
	ID                 int64           `gorm:"column:id;primaryKey" json:"id"`
	RunID              int64           `gorm:"column:run_id;not null" json:"runId"`
	RoundNumber        int             `gorm:"column:round_number;not null" json:"roundNumber"`
	Status             string          `gorm:"column:status;size:30;not null" json:"status"`
	InputHypotheses    json.RawMessage `gorm:"column:input_hypotheses;type:jsonb;not null" json:"inputHypotheses"`
	NewEvidenceIDs     json.RawMessage `gorm:"column:new_evidence_ids;type:jsonb;not null" json:"newEvidenceIds"`
	RejectedHypotheses json.RawMessage `gorm:"column:rejected_hypotheses;type:jsonb;not null" json:"rejectedHypotheses"`
	NextActions        json.RawMessage `gorm:"column:next_actions;type:jsonb;not null" json:"nextActions"`
	ErrorCode          *string         `gorm:"column:error_code;size:80" json:"errorCode,omitempty"`
	StartedAt          *time.Time      `gorm:"column:started_at" json:"startedAt,omitempty"`
	FinishedAt         *time.Time      `gorm:"column:finished_at" json:"finishedAt,omitempty"`
	CreatedAt          time.Time       `gorm:"column:created_at;autoCreateTime" json:"createdAt"`
	UpdatedAt          time.Time       `gorm:"column:updated_at;autoUpdateTime" json:"updatedAt"`
}

func (RCARound) TableName() string { return "rca_round" }

type RCAAction struct {
	ID            int64           `gorm:"column:id;primaryKey" json:"id"`
	RunID         int64           `gorm:"column:run_id;not null" json:"runId"`
	RoundID       int64           `gorm:"column:round_id;not null" json:"roundId"`
	ActionKey     string          `gorm:"column:action_key;size:160;not null" json:"actionKey"`
	SkillName     string          `gorm:"column:skill_name;size:120;not null" json:"skillName"`
	Status        string          `gorm:"column:status;size:30;not null" json:"status"`
	Input         json.RawMessage `gorm:"column:input;type:jsonb;not null" json:"input"`
	Output        json.RawMessage `gorm:"column:output;type:jsonb" json:"output,omitempty"`
	EvidenceIDs   json.RawMessage `gorm:"column:evidence_ids;type:jsonb;not null" json:"evidenceIds"`
	ErrorCode     *string         `gorm:"column:error_code;size:80" json:"errorCode,omitempty"`
	ErrorMessage  *string         `gorm:"column:error_message" json:"errorMessage,omitempty"`
	SensitiveRead bool            `gorm:"column:sensitive_read;not null" json:"sensitiveRead"`
	Attempt       int             `gorm:"column:attempt;not null" json:"attempt"`
	StartedAt     *time.Time      `gorm:"column:started_at" json:"startedAt,omitempty"`
	FinishedAt    *time.Time      `gorm:"column:finished_at" json:"finishedAt,omitempty"`
	CreatedAt     time.Time       `gorm:"column:created_at;autoCreateTime" json:"createdAt"`
	UpdatedAt     time.Time       `gorm:"column:updated_at;autoUpdateTime" json:"updatedAt"`
}

func (RCAAction) TableName() string { return "rca_action" }

type RCARootCauseCandidate struct {
	ID          int64           `gorm:"column:id;primaryKey" json:"id"`
	RunID       int64           `gorm:"column:run_id;not null" json:"runId"`
	RoundID     int64           `gorm:"column:round_id;not null" json:"roundId"`
	Summary     string          `gorm:"column:summary;not null" json:"summary"`
	Confidence  float64         `gorm:"column:confidence;not null" json:"confidence"`
	EvidenceIDs json.RawMessage `gorm:"column:evidence_ids;type:jsonb;not null" json:"evidenceIds"`
	Rejected    bool            `gorm:"column:rejected;not null" json:"rejected"`
	CreatedAt   time.Time       `gorm:"column:created_at;autoCreateTime" json:"createdAt"`
	UpdatedAt   time.Time       `gorm:"column:updated_at;autoUpdateTime" json:"updatedAt"`
}

func (RCARootCauseCandidate) TableName() string { return "rca_root_cause_candidate" }
